provider/terraform_kcc.rb

```diff
@@ -0,0 +1,133 @@
+      examples = data.object.examples
+                     .reject(&:skip_test)
+                     .reject { |e| !e.test_env_vars.nil? && e.test_env_vars.any? }
+                     .reject { |e| @version < @api.version_obj_or_closest(e.min_version) }
+
+      examples.each do |example|
+        target_folder = File.join(data.output_folder, 'samples', example.name)
+        FileUtils.mkpath target_folder
+
+        data.example = example
+
+        data.generate(
+          'templates/kcc/samples/sample.tf.erb',
+          File.join(target_folder, 'main.tf'),
+          self
+        )
+      end
+    end
+
+    def generate_resource_tests(data) end
+
+    def generate_iam_policy(data) end
+
+    def compile_product_files(output_folder)
+      file_template = ProductFileTemplate.new(
+        output_folder,
+        nil,
+        @api,
+        @target_version_name,
+        build_env
+      )
+      compile_file_list(output_folder,
+                        [
+                          [
+                            "servicemappings/#{@api.name.downcase}.yaml",
+                            'templates/kcc/product/service_mapping.yaml.erb'
+                          ]
+                        ],
+                        file_template)
+    end
+
+    def compile_common_files(output_folder, products, _common_compile_file)
+      Google::LOGGER.info 'Compiling common files.'
+      file_template = ProviderFileTemplate.new(
+        output_folder,
+        @target_version_name,
+        build_env,
+        products
+      )
+      compile_file_list(output_folder, [
+                          [
+                            'common/resources.go',
+                            'templates/kcc/controller_resources.go.erb'
+                          ]
+                        ], file_template)
+    end
+
+    def copy_common_files(output_folder)
+      Google::LOGGER.info 'Copying common files.'
+      copy_file_list(output_folder, [])
+    end
+
+    # A strict mapping from K8S name -> Terraform resource "name" doesn't make
+    # sense for some resources but we can approximate this well enough for most
+    # of them. This is often the `name` field, but it could be named something
+    # else.
+    # If {{name}} is part of a resource id, it should be the last import format.
+    # Otherwise, {{value}} or values/{{value}} are also valid. If the final id
+    # has multiple terms, we can reject it as we can't create a 1:1 mapping from
+    # K8S resource name : Terraform field.
+    def guess_metadata_mapping_name(object)
+      final_import_parts = import_id_formats(object)[-1].scan(/{{[[:word:]]+}}/)
+
+      # remove {{project}}, {{region}} if present, KCC handles them specially
+      # {{location}} is (generally) a region, so it can be removed as well.
+      final_import_parts -= ['{{project}}', '{{region}}', '{{location}}']
+
+      final_import_parts.first.gsub('{{', '').gsub('}}', '') if final_import_parts.length == 1
```

Comment:
This will return nil if final_import_parts.length is 0 or > 1. Is that expected or should it throw an error?

(It looks like the template below will ignore name altogether)

Reply:
That's expected, https://github.com/GoogleCloudPlatform/magic-modules/pull/2279/files#diff-c4618627ddf17fbf071cf0add98b9660R69 performed a nil check. Also updated the header to say it returns nil if we don't find a 1:1 mapping.